cirq/sim/simulator.py

```diff
@@ -512,17 +515,75 @@ def _base_iterator(
     ) -> Iterator[TStepResult]:
         """Iterator over StepResult from Moments of a Circuit.
 
+        This is a thin wrapper around `create_act_on_args` and `_core_iterator`.
+        Overriding this method was the old way of creating a circuit iterator,
+        and this method is planned to be formally put on the deprecation path.
+        Going forward, override the aforementioned two methods in custom
+        simulators.
+
         Args:
             circuit: The circuit to simulate.
-            param_resolver: A ParamResolver for determining values of
-                Symbols.
             qubit_order: Determines the canonical ordering of the qubits. This
                 is often used in specifying the initial state, i.e. the
                 ordering of the computational basis states.
             initial_state: The initial state for the simulation. The form of
                 this state depends on the simulation implementation. See
                 documentation of the implementing class for details.
 
+        Yields:
+            StepResults from simulating a Moment of the Circuit.
+        """
+        qubits = ops.QubitOrder.as_qubit_order(qubit_order).order_for(circuit.all_qubits())
+        act_on_args = (
+            cast(TActOnArgs, initial_state)
+            if isinstance(initial_state, ActOnArgs)
+            else self.create_act_on_args(initial_state, qubits)
+        )
+        return self._core_iterator(circuit, act_on_args, qubits)
+
+    @abc.abstractmethod
+    def create_act_on_args(
+        self,
+        initial_state: Any,
+        qubits: Tuple['cirq.Qid', ...],
+    ) -> TActOnArgs:
+        """Creates the ActOnArgs state for a simulator.
+
+        Custom simulators should implement this method.
+
+        Args:
+            initial_state: The initial state for the simulation. The form of
+                this state depends on the simulation implementation. See
+                documentation of the implementing class for details.
+            qubits: Determines the canonical ordering of the qubits. This
+                is often used in specifying the initial state, i.e. the
+                ordering of the computational basis states.
+
+        Returns:
+            The ActOnArgs for this simulator.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _core_iterator(
+        self,
+        circuit: circuits.Circuit,
+        initial_state: TActOnArgs,
+        qubits: Tuple['cirq.Qid', ...],
```

Comment:
Consider putting this and qubit_map into ActOnArgs itself, to avoid having to pass this parameter around everywhere with it.

Reply:
Done but let's write a unit test to make sure it works as desired.

Reply:
Tried this but it explodes. Try again later.